### PR TITLE
Lightning refactor losses

### PR DESCRIPTION
**Pull request by szmazurek**
Fixes #

## Proposed Changes
- extract loss calculation logic to separate classes 
- add get_ methods for model, loss, scheduler
- move parameter validation logic to the get_functions

## Checklist

- [ ] [`CONTRIBUTING`](https://github.com/mlcommons/GaNDLF/blob/master/CONTRIBUTING.md) guide has been followed.
- [ ] PR is based on the [current GaNDLF master ](https://docs.github.com/en/desktop/contributing-and-collaborating-using-github-desktop/keeping-your-local-repository-in-sync-with-github/syncing-your-branch-in-github-desktop?platform=windows).
- [ ] Non-breaking change (does **not** break existing functionality): provide **as many** details as possible for _any_ breaking change.
- [ ] Function/class source code documentation added/updated (ensure `typing` is used to provide type hints, including and not limited to using `Optional` if a variable has a pre-defined value).
- [ ] Code has been [blacked](https://github.com/psf/black#usage) for style consistency and linting.
- [ ] If applicable, version information [has been updated in GANDLF/version.py](https://github.com/mlcommons/GaNDLF/blob/master/GANDLF/version.py).
- [ ] If adding a git submodule, add to list of exceptions for black styling in [pyproject.toml](https://github.com/mlcommons/GaNDLF/blob/master/pyproject.toml) file.
- [ ] [Usage documentation](https://github.com/mlcommons/GaNDLF/blob/master/docs) has been updated, if appropriate.
- [ ] Tests added or modified to [cover the changes](https://app.codecov.io/gh/mlcommons/GaNDLF); if coverage is reduced, please give explanation.
- [ ] If customized dependency installation is required (i.e., a separate `pip install` step is needed for PR to be functional), please ensure it is reflected in all the files that control the CI, namely: [python-test.yml](https://github.com/mlcommons/GaNDLF/blob/master/.github/workflows/python-test.yml), and all docker files [[1](https://github.com/mlcommons/GaNDLF/blob/master/Dockerfile-CPU),[2](https://github.com/mlcommons/GaNDLF/blob/devcontainer_build_fix/Dockerfile-CUDA11.6),[3](https://github.com/mlcommons/GaNDLF/blob/master/Dockerfile-ROCm)].
- [ ] The `logging` library is being used and no `print` statements are left.

### CODE REVIEW

File: GANDLF/losses/__init__.py [hunk 1]
```diff
@@ -38,3 +38,26 @@
     "focal": FocalLoss,
     "dc_focal": DC_Focal,
 }
+
+
+def get_loss(params):
```
Comment (sarthakpati) [inline GANDLF/losses/__init__.py]: What should the return type be?

```suggestion
def get_loss(params: dict):
```

File: GANDLF/losses/loss_calculators.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
+import torch
+import torch.nn.functional as F
+from GANDLF.losses import get_loss
+from GANDLF.utils.tensor import reverse_one_hot, get_linear_interpolation_mode
+from abc import ABC, abstractmethod
+from typing import List
+
+
+class AbstractLossCalculator(ABC):
```
Comment (sarthakpati) [inline GANDLF/losses/loss_calculators.py:9]: Take a look at the comment in this issue as well: https://github.com/mlcommons/GaNDLF/issues/854

File: GANDLF/models/lightning_module.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+from GANDLF.models import get_model
+from GANDLF.optimizers import get_optimizer
+from GANDLF.schedulers import get_scheduler
+from GANDLF.losses.loss_calculators import LossCalculatorFactory
+from GANDLF.models.modelBase import ModelBase
+
+from copy import deepcopy
+
+
+class GandlfLightningModule(pl.LightningModule):
```
Comment (sarthakpati) [inline GANDLF/models/lightning_module.py:11]: If this is part of the model base, should this perhaps not be in the same module?

File: GANDLF/losses/__init__.py (hunk 1 above)
Comment (szmazurek) [reply GANDLF/losses/__init__.py]: Will add

File: GANDLF/losses/loss_calculators.py (hunk 2 above)
Comment (szmazurek) [reply GANDLF/losses/loss_calculators.py:9]: I did, and I also agree with Slava. Packing the losses into classes that implement the __call__ method would basically let us keep the existing codebase nearly intact while giving more organization to the code. Should I do that in the meantime? I can open a separate PR.

And regarding this abstract interface - I think we will still need it, as it addresses some specific combinations and operations that are related to the model, so using the losses but in a specific way. What do you think?

File: GANDLF/models/lightning_module.py (hunk 3 above)
Comment (szmazurek) [reply GANDLF/models/lightning_module.py:11]: We may move it, but I treat the model base as a base class for the architectures, and the lighting module is kind of orchestrator of not only architecture but also the training, inference loops etc. So I would basically say it is a higher level of abstraction, thus I put it in a separate file. Also this will never be subclassed I think.

File: GANDLF/losses/loss_calculators.py (hunk 2 above)
Comment (szmazurek) [reply GANDLF/losses/loss_calculators.py:9]: Btw regarding those - what do you think, if we decide to implement the common interface for losses, that the params are an instance variable set in the initialization and the call only uses necessary inputs? We could hide this initialization behind the get_loss method, so once again from perspective of codebase it will be invisible

Comment (sarthakpati) [reply GANDLF/losses/loss_calculators.py:9]: Completely agree. 👍🏽

File: GANDLF/models/lightning_module.py (hunk 3 above)
Comment (sarthakpati) [reply GANDLF/models/lightning_module.py:11]: Ah okay, I understand, thank you!

File: GANDLF/losses/loss_calculators.py (hunk 2 above)
Comment (szmazurek) [reply GANDLF/losses/loss_calculators.py:9]: noted, will create pr for that

**Comment by szmazurek** [comment]
@sarthakpati can I merge?
